lib/Dialect/LLHD/Transforms/InlineCalls.cpp

```diff
@@ -0,0 +1,164 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/LLHD/IR/LLHDOps.h"
+#include "circt/Dialect/LLHD/Transforms/LLHDPasses.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Threading.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Interfaces/CallInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/Inliner.h"
+#include "mlir/Transforms/InliningUtils.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "llhd-inline-calls"
+
+namespace circt {
+namespace llhd {
+#define GEN_PASS_DEF_INLINECALLSPASS
+#include "circt/Dialect/LLHD/Transforms/LLHDPasses.h.inc"
+} // namespace llhd
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+using namespace llhd;
+using llvm::SmallSetVector;
+
+namespace {
+/// Implementation of the `InlinerInterface` that allows calls in SSACFG regions
+/// nested within `llhd.process`, `llhd.final`, and `llhd.combinational` ops to
+/// be inlined.
+struct FunctionInliner : public InlinerInterface {
+  using InlinerInterface::InlinerInterface;
+
+  bool isLegalToInline(Operation *call, Operation *callable,
+                       bool wouldBeCloned) const override {
+    // Only inline `func.func` ops.
+    if (!isa<func::FuncOp>(callable))
+      return false;
+
+    // Only inline into SSACFG regions embedded within LLHD processes.
+    if (!mayHaveSSADominance(*call->getParentRegion()))
+      return false;
+    auto *op = call;
+    while ((op = op->getParentOp()))
+      if (isa<ProcessOp, FinalOp, CombinationalOp>(op))
+        return true;
+    return false;
+  }
+
+  bool isLegalToInline(Region *dest, Region *src, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    return true;
+  }
+
+  bool isLegalToInline(Operation *op, Region *dest, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    return true;
+  }
+
+  bool shouldAnalyzeRecursively(Operation *op) const override { return false; }
+};
+
+/// Pass implementation.
+struct InlineCallsPass
+    : public llhd::impl::InlineCallsPassBase<InlineCallsPass> {
+  using CallStack = SmallSetVector<func::FuncOp, 8>;
+  void runOnOperation() override;
+  void runOnRegion(Region &region, const SymbolTable &symbolTable,
+                   CallStack &callStack);
+};
+} // namespace
+
+void InlineCallsPass::runOnOperation() {
+  auto &symbolTable = getAnalysis<SymbolTable>();
+  parallelForEach(&getContext(), getOperation().getOps<hw::HWModuleOp>(),
+                  [&](auto module) {
+                    CallStack callStack;
+                    runOnRegion(module.getBody(), symbolTable, callStack);
+                  });
```

Comment:
Can we change runOnRegion to return LogicalResult and use `parallelForEachFailable` here?